### ISSUE TITLE
agate-sql: csvsql: Specify precision and scale for DECIMAL

### ISSUE THREAD

**Issue by jkinable** [wireservice/csvkit#856]
Given the following table:
```
ID | X | Ref_ID
1 | 42.4242424242 | 99900000002
```

Invoking `csvsql --dialect mysql --snifflimit 100000 datatwithheaders.csv > mytabledef.sql` on this table produces:
```
CREATE TABLE `Facility` (
	`ID` DECIMAL NOT NULL, 
	`X` DECIMAL, 
	`Ref_ID` DECIMAL NOT NULL, 
);
```

When importing the data into the Facility table, the values unexpectedly change to:
1 | 42 | 9999999999

This is because the create table function produced by csvsql neglected to specify the precision and scale of the decimals:
- 42.4242424242 changes to 42 because the type of this row is DECIMAL(10,0). Hence the decimal values are truncated. In this case, X is a geographic coordinate, so this column should have been created with DECIMAL(12,10)
- 99900000002 changes to 9999999999 because DECIMAL(10,0) is not large enough to hold this number.

**Comment by jpmckinney** [comment]
So the bug is in agate-sql but I'm less familiar with that code: https://github.com/wireservice/agate-sql/blob/master/agatesql/table.py

**Comment by clarsenfvt** [comment]
I'm having this same issue. I can get around it by disabling the inference with -I, but that results in everything being stored as a varchar, which isn't optimal. If I use inference, all my numbers (which originally have values like 3.696700 all get truncated to zero decimal places (so 3 in this case).

The command I'm using is csvsql --db {REDACTED} --insert --overwrite --tables testTable home/../data.csv

**Comment by jkinable** [comment]
This issue is relatively easy to fix, but does take some work and careful testing. The issue is caused by the following bits of code in https://github.com/wireservice/agate-sql/blob/master/agatesql/table.py:

1. 
```
if py_type in [int, float, decimal.Decimal]:
            if py_type is float:
                sql_column.type.asdecimal = True
            column_types.append(agate.Number())
```

2. 
```
SQL_TYPE_MAP = {
    agate.Boolean: BOOLEAN,
    agate.Number: DECIMAL,
    agate.Date: DATE,
    agate.DateTime: TIMESTAMP,
    agate.TimeDelta: None,  # See below
    agate.Text: VARCHAR
}
```
This is what happens: 
Data types are for SQL are queried from the `SQL_TYPE_MAP`. In this particular case, if the data type of a column is int, float or decimal.Decimal, the agate.Number key is used which corresponds to the `DECIMAL `type. In other words, a new SQL column of the type DECIMAL is created. According to the MYSQL [manual](https://dev.mysql.com/doc/refman/5.5/en/fixed-point-types.html), `DECIMAL` is short for `DECIMAL(10,0)`. This datatype can store ten digits and zero decimals! So the largest integer this data type can hold is 9999999999, and this data type cannot hold any fractional values! 

What should have been done:
1. Distinguish between integer and fractional values (i.e. split the if statement in my first code snippet).
2a. In case of integer, scan the entire column to find the largest integer and determine the smallest data type that can exactly represent this integer: `TINYINT`, `SMALLINT`, `MEDIUMINT`, `INT`, `BIGINT`
Remember to distinguish between signed and unsigned integers.
2b. In case of fractional values, use the `DECIMAL` type. Again scan the entire column and check (a) the largest number of digits m, and the largest number of decimals n. Next, create a column DECIMAL(m,n). Do not use any of the approximate data types such as `FLOAT`, `DOUBLE`

**Comment by jpmckinney** [comment]
@jkinable A special challenge is that we don't want to have to have a different map for every SQL implementation, because not all implementations offer the same types. (For example, PostgreSQL has no `DATETIME`. SQLite and others are simplified.)

I'm fixing in agate-sql https://github.com/wireservice/agate-sql/pull/26, so should be fixed once I release agate-sql and a new version of csvkit. Note that PostgreSQL and SQLite don't require specifying a decimal's precision and scale.

**Comment by codejunkie94** [comment]
@jpmckinney Facing the same issue with snowflake dialect as it also requires specifying the scale and precision for `DECIMAL` type. Can you add a fix for snowflake similar to the change [wireservice/agate-sql#26](url)?

**Comment by jpmckinney** [comment]
Can you please open a new issue? I don't know snowflake.

**Comment by mvl22** [comment]
Note to anyone running into this who is using Ubuntu [20.04 (focal)](https://packages.ubuntu.com/focal/) - the release there is at 1.0.2, but the fix noted above went into 1.0.3.
